File: src/lib/tauri-mock/__tests__/conformance-snapshot.ts
/**
 * #763 — the shared normalized-snapshot builder (TS / tauri-mock side).
 *
 * This is the TS twin of the Rust snapshot builder at
 * `src-tauri/src/command_integration_tests/conformance_snapshot.rs`. It must
 * produce a snapshot that is BYTE-IDENTICAL (after canonical key sorting) to
 * the one the backend authors for the same logical state. Keep the two in
 * lockstep — the canonical id relabeling, field projection, sort order, and
 * op_log-digest normalization rules are duplicated deliberately so an identical
 * state on both sides yields identical JSON.
 *
 * See the Rust module's doc comment for the relabeling contract.
 */

// ---------------------------------------------------------------------------
// Mock state shapes (subset of what handlers.ts reads/writes).
// ---------------------------------------------------------------------------

type MockBlock = Record<string, unknown>
type MockPropRow = Record<string, unknown>

export interface MockState {
  blocks: Map<string, MockBlock>
  properties: Map<string, Map<string, MockPropRow>>
  blockTags: Map<string, Set<string>>
  /** Ordered op_log: each entry has at least `op_type` and a JSON `payload`. */
  opLog: ReadonlyArray<{ op_type: string; payload: string }>
}

// ---------------------------------------------------------------------------
// Normalized snapshot shape (mirror of the Rust `Snapshot`).
// ---------------------------------------------------------------------------

export interface NormalizedSnapshot {
  blocks: Array<Record<string, unknown>>
  properties: Array<Record<string, unknown>>
  block_tags: Array<Record<string, unknown>>
  page_links: Array<Record<string, unknown>>
  op_log_digest: { count: number; ops: Array<Record<string, unknown>> }
}

// The same `[[ULID]]` link regex the mock's `deriveLinkEdges` uses. The mock
// derives page links live from non-deleted block content; we replicate that
// here so the TS snapshot is symmetric with the backend's `block_links` table.
const LINK_RE = /\[\[([0-9A-Z]{26})\]\]/g

/** Numeric sort key for a `Bn` canonical token (`B12` < `B2` would be wrong
 *  lexically). Pass-through (non-`Bn`) tokens sort after all `Bn` tokens. */
function tokenKey(token: string): [number, number, string] {
  const m = /^B(\d+)$/.exec(token)
  if (m) return [0, Number(m[1]), '']
  return [1, 0, token]
}

function cmpTokens(a: string, b: string): number {
  const [ap, an, as_] = tokenKey(a)
  const [bp, bn, bs] = tokenKey(b)
  if (ap !== bp) return ap - bp
  if (an !== bn) return an - bn
  return as_ < bs ? -1 : as_ > bs ? 1 : 0
}

/**
 * Build the canonical id → `Bn` token map from the ordered list of raw ids
 * (seed ids in seed order, then created ids in op order — computed by the
 * caller from the fixture seed + the mock op_log's create_block entries).
 */
export function canonicalLabelMap(canonicalOrder: ReadonlyArray<string>): Map<string, string> {
  const map = new Map<string, string>()
  canonicalOrder.forEach((id, i) => {
    map.set(id, `B${i + 1}`)
  })
  return map
}

/**
 * Canonicalize a raw mock op_log entry into a digest entry (or `null` to drop
 * it). Mirrors `RawOp::canonicalize` on the Rust side: auto-derived timestamp
 * property writes are dropped; reserved-key set_property ops collapse to their
 * `set_<key>` logical name. The mock already emits dedicated op_types
 * (`set_todo_state`, …) so most entries pass through unchanged.
 */
function canonicalizeOp(opType: string, key: string | null): Record<string, unknown> | null {
  if (opType === 'set_property' && (key === 'created_at' || key === 'completed_at')) {
    return null
  }
  if (
    opType === 'set_property' &&
    key != null &&
    (key === 'todo_state' || key === 'priority' || key === 'due_date' || key === 'scheduled_date')
  ) {
    return { op_type: `set_${key}` }
  }
  if (opType === 'set_property') {
    return key != null ? { op_type: 'set_property', key } : { op_type: 'set_property' }
  }
  return { op_type: opType }
}

/**
 * Derive the `value_type` + `value` for a property row from whichever typed
 * column is set. Ref values are relabeled. Mirrors `property_typed_value`.
 */
function propertyTypedValue(
  row: MockPropRow,
  relabel: (id: string) => string,
): { value_type: string | null; value: unknown } {
  const text = row['value_text'] as string | null | undefined
  const num = row['value_num'] as number | null | undefined
  const date = row['value_date'] as string | null | undefined
  const ref = row['value_ref'] as string | null | undefined
  const bool = row['value_bool'] as number | boolean | null | undefined
  if (text != null) return { value_type: 'Text', value: text }
  if (num != null) return { value_type: 'Num', value: num }
  if (date != null) return { value_type: 'Date', value: date }
  if (ref != null) return { value_type: 'Ref', value: relabel(ref) }
  if (bool != null) return { value_type: 'Bool', value: bool === true || bool === 1 }
  return { value_type: null, value: null }
}

/**
 * Build the full normalized snapshot from the live mock state. `canonicalOrder`
 * supplies the relabel order (identical computation to the Rust runner).
 */
export function buildSnapshot(
  state: MockState,
  canonicalOrder: ReadonlyArray<string>,
): NormalizedSnapshot {
  const map = canonicalLabelMap(canonicalOrder)
  const relabel = (id: string): string => map.get(id) ?? id
  const relabelOpt = (id: string | null | undefined): string | null =>
    id == null ? null : relabel(id)

  // Blocks — sorted by canonical id.
  const blocks = [...state.blocks.values()]
    .map((b) => ({
      id: relabel(b['id'] as string),
      block_type: b['block_type'] as string,
      content: (b['content'] as string | null | undefined) ?? null,
      parent_id: relabelOpt(b['parent_id'] as string | null | undefined),
      page_id: relabelOpt(b['page_id'] as string | null | undefined),
      position: (b['position'] as number | null | undefined) ?? null,
      deleted_at: b['deleted_at'] == null ? null : 'DELETED',
      todo_state: (b['todo_state'] as string | null | undefined) ?? null,
      priority: (b['priority'] as string | null | undefined) ?? null,
      due_date: (b['due_date'] as string | null | undefined) ?? null,
      scheduled_date: (b['scheduled_date'] as string | null | undefined) ?? null,
    }))
    .sort((a, b) => cmpTokens(a.id, b.id))

  // Properties — block_properties projection. Exclude auto-derived timestamp
  // keys (created_at/completed_at) to match the Rust side.
  const propertyRows: Array<Record<string, unknown>> = []
  for (const [blockId, keyMap] of state.properties.entries()) {
    for (const [key, row] of keyMap.entries()) {
      if (key === 'created_at' || key === 'completed_at') continue
      // The mock stamps a `space` ref property on pages; the backend keeps
      // `space` in the column-backed `blocks.space_id` (NOT block_properties),
      // so it never appears in the snapshot's `properties`. Skip it here too.
      if (key === 'space') continue
      const { value_type, value } = propertyTypedValue(row, relabel)
      propertyRows.push({
        block_id: relabel(blockId),
        key,
        value_type,
        value,
      })
    }
  }
  const properties = propertyRows.sort((a, b) => {
    const c = cmpTokens(a['block_id'] as string, b['block_id'] as string)
    if (c !== 0) return c
    const ak = a['key'] as string
    const bk = b['key'] as string
    return ak < bk ? -1 : ak > bk ? 1 : 0
  })

  // Block tags — sorted by (block_id, tag_id).
  const tagRows: Array<Record<string, unknown>> = []
  for (const [blockId, tagSet] of state.blockTags.entries()) {
    for (const tagId of tagSet) {
      tagRows.push({ block_id: relabel(blockId), tag_id: relabel(tagId) })
    }
  }
  const block_tags = tagRows.sort((a, b) => {
    const c = cmpTokens(a['block_id'] as string, b['block_id'] as string)
    if (c !== 0) return c
    return cmpTokens(a['tag_id'] as string, b['tag_id'] as string)
  })

  // Page links — derive [[ULID]] edges from non-deleted block content, mirror
  // of the backend's `block_links` table joined to the source's page_id. The
  // target must reference a live block (mirrors the backend's EXISTS guard).
  const liveIds = new Set(
    [...state.blocks.values()].filter((b) => b['deleted_at'] == null).map((b) => b['id'] as string),
  )
  const linkRows: Array<Record<string, unknown>> = []
  for (const b of state.blocks.values()) {
    if (b['deleted_at'] != null) continue
    const content = (b['content'] as string | null) ?? ''
    if (!content.includes('[[')) continue
    const sourceId = b['id'] as string
    const sourcePageId = (b['page_id'] as string | null) ?? null
    LINK_RE.lastIndex = 0
    for (const m of content.matchAll(LINK_RE)) {
      const targetId = m[1] as string
      if (!liveIds.has(targetId)) continue
      linkRows.push({
        source_id: relabel(sourceId),
        target_id: relabel(targetId),
        source_page_id: relabelOpt(sourcePageId),
      })
    }
  }
  const page_links = linkRows.sort((a, b) => {
    let c = cmpTokens(a['source_id'] as string, b['source_id'] as string)
    if (c !== 0) return c
    c = cmpTokens(a['target_id'] as string, b['target_id'] as string)
    if (c !== 0) return c
    return cmpTokens(
      (a['source_page_id'] as string | null) ?? '',
      (b['source_page_id'] as string | null) ?? '',
    )
  })

  // Op log digest — ordered (the mock appends in op order), filtered/canonicalized.
  const ops: Array<Record<string, unknown>> = []
  for (const entry of state.opLog) {
    let key: string | null = null
    try {
      const payload = JSON.parse(entry.payload) as Record<string, unknown>
      key = (payload['key'] as string | undefined) ?? null
    } catch {
      key = null
    }
    const digest = canonicalizeOp(entry.op_type, key)
    if (digest != null) ops.push(digest)
  }
  const op_log_digest = { count: ops.length, ops }

  return { blocks, properties, block_tags, page_links, op_log_digest }
}

File: src-tauri/src/command_integration_tests/conformance_snapshot.rs
//! #763 — the shared normalized-snapshot builder (Rust side).
//!
//! This is the crux of the conformance harness: it must produce a snapshot
//! that is BYTE-IDENTICAL to the one the TS side builds from the tauri-mock
//! for the same logical state. The TS twin lives at
//! `src/lib/tauri-mock/__tests__/conformance-snapshot.ts` — keep the two in
//! lockstep.
//!
//! ## Canonical id relabeling
//!
//! Block ids differ per backend (ULID) vs mock (counter), so the raw id value
//! is useless for comparison. Instead we relabel every id to a canonical token
//! `B1, B2, …` by a DETERMINISTIC traversal that does not depend on the raw id:
//!
//!   * seed blocks first, in fixture seed order;
//!   * then created blocks, in op (creation) order — i.e. the order their
//!     `create_block` op was appended to the op_log.
//!
//! Every id REFERENCE is remapped through the same map: `parent_id`, `page_id`,
//! `block_tags.{block_id,tag_id}`, `page_links.{source_id,target_id,source_page_id}`,
//! and property `value_ref`. A reference to a block that is not in the map
//! (should not happen for in-corpus fixtures) is passed through verbatim so a
//! drift shows up rather than panicking.
//!
//! Nondeterministic fields (absolute timestamps, hashes, cursors, raw ids) are
//! stripped. `deleted_at` collapses to a boolean-ish sentinel: `null` when the
//! block is live, `"DELETED"` when tombstoned.

use serde::Serialize;
use serde_json::{Value, json};
use std::collections::BTreeMap;

// ---------------------------------------------------------------------------
// Raw (pre-relabel) intermediate state, filled by the runner's DB reads.
// ---------------------------------------------------------------------------

pub struct RawBlock {
    pub id: String,
    pub block_type: String,
    pub content: Option<String>,
    pub parent_id: Option<String>,
    pub position: Option<i64>,
    pub deleted: bool,
    pub todo_state: Option<String>,
    pub priority: Option<String>,
    pub due_date: Option<String>,
    pub scheduled_date: Option<String>,
    pub page_id: Option<String>,
}

pub struct RawProperty {
    pub block_id: String,
    pub key: String,
    pub value_text: Option<String>,
    pub value_num: Option<f64>,
    pub value_date: Option<String>,
    pub value_ref: Option<String>,
    pub value_bool: Option<bool>,
}

pub struct RawTag {
    pub block_id: String,
    pub tag_id: String,
}

pub struct RawLink {
    pub source_id: String,
    pub target_id: String,
    pub source_page_id: Option<String>,
}

/// A digest entry: the canonicalized op type and, for property ops, the key.
pub struct RawOp {
    pub op_type: String,
    pub key: Option<String>,
}

impl RawOp {
    /// Canonicalize a raw op_log row into a digest entry, applying the shared
    /// normalization rules. Returns `None` for ops that are dropped from the
    /// digest (auto-derived timestamp property writes).
    pub fn canonicalize(op_type: &str, key: Option<&str>) -> Option<Self> {
        // Auto-derived timestamp property writes — the mock never models these,
        // and they carry today's date. Drop from the digest.
        if op_type == "set_property" && matches!(key, Some("created_at" | "completed_at")) {
            return None;
        }
        // Reserved-key set_property ops are emitted by the backend as
        // `set_property(key=<reserved>)` but the mock has dedicated op_types
        // (`set_todo_state`, `set_due_date`, …). Canonicalize the backend shape
        // to the mock's logical name so the digests line up.
        if op_type == "set_property"
            && let Some(k) = key
            && matches!(k, "todo_state" | "priority" | "due_date" | "scheduled_date")
        {
            return Some(RawOp {
                op_type: format!("set_{k}"),
                key: None,
            });
        }
        let key = if op_type == "set_property" {
            key.map(str::to_owned)
        } else {
            None
        };
        Some(RawOp {
            op_type: op_type.to_owned(),
            key,
        })
    }
}

pub struct RawState {
    pub blocks: Vec<RawBlock>,
    pub properties: Vec<RawProperty>,
    pub block_tags: Vec<RawTag>,
    pub page_links: Vec<RawLink>,
    pub op_log: Vec<RawOp>,
}

// ---------------------------------------------------------------------------
// Normalized snapshot (the serialized comparison target).
// ---------------------------------------------------------------------------

/// The normalized snapshot. Serialized to JSON and compared field-by-field via
/// a `BTreeMap` round-trip on both sides, so map-key order never matters; the
/// arrays below ARE order-sensitive and are sorted deterministically here.
#[derive(Serialize)]
pub struct Snapshot {
    pub blocks: Vec<Value>,
    pub properties: Vec<Value>,
    pub block_tags: Vec<Value>,
    pub page_links: Vec<Value>,
    pub op_log_digest: Value,
}

/// Build the canonical id → token (`B1`, `B2`, …) map from the ordered list of
/// raw ids. The order is computed by the runner: seed ids in seed order, then
/// created ids in op order.
pub fn canonical_label_map(canonical_order: &[String]) -> BTreeMap<String, String> {
    canonical_order
        .iter()
        .enumerate()
        .map(|(i, id)| (id.clone(), format!("B{}", i + 1)))
        .collect()
}

/// Build the full normalized snapshot. Relabeling is driven entirely by
/// `canonical_order` (seed ids in seed order, then created ids in op order).
pub fn build_snapshot_with_order(state: RawState, canonical_order: &[String]) -> Snapshot {
    let map = canonical_label_map(canonical_order);
    let relabel = |id: &str| -> String { map.get(id).cloned().unwrap_or_else(|| id.to_owned()) };
    let relabel_opt = |id: &Option<String>| -> Value {
        match id {
            Some(s) => json!(relabel(s)),
            None => Value::Null,
        }
    };

    // Blocks — sorted by canonical id (numeric on the Bn suffix).
    let mut blocks: Vec<(String, Value)> = state
        .blocks
        .into_iter()
        .map(|b| {
            let cid = relabel(&b.id);
            let v = json!({
                "id": cid,
                "block_type": b.block_type,
                "content": b.content,
                "parent_id": relabel_opt(&b.parent_id),
                "page_id": relabel_opt(&b.page_id),
                "position": b.position,
                "deleted_at": if b.deleted { json!("DELETED") } else { Value::Null },
                "todo_state": b.todo_state,
                "priority": b.priority,
                "due_date": b.due_date,
                "scheduled_date": b.scheduled_date,
            });
            (cid, v)
        })
        .collect();
    blocks.sort_by_key(|(cid, _)| token_key(cid));
    let blocks: Vec<Value> = blocks.into_iter().map(|(_, v)| v).collect();

    // Properties — sorted by (block_id, key).
    let mut properties: Vec<(String, String, Value)> = state
        .properties
        .into_iter()
        .map(|p| {
            let bid = relabel(&p.block_id);
            let (value_type, value) = property_typed_value(&p, &relabel);
            let v = json!({
                "block_id": bid,
                "key": p.key,
                "value_type": value_type,
                "value": value,
            });
            (bid, p.key, v)
        })
        .collect();
    properties.sort_by(|a, b| (token_key(&a.0), &a.1).cmp(&(token_key(&b.0), &b.1)));
    let properties: Vec<Value> = properties.into_iter().map(|(_, _, v)| v).collect();

    // Block tags — sorted by (block_id, tag_id).
    let mut block_tags: Vec<(String, String, Value)> = state
        .block_tags
        .into_iter()
        .map(|t| {
            let bid = relabel(&t.block_id);
            let tid = relabel(&t.tag_id);
            (
                bid.clone(),
                tid.clone(),
                json!({ "block_id": bid, "tag_id": tid }),
            )
        })
        .collect();
    block_tags.sort_by_key(|(bid, tid, _)| (token_key(bid), token_key(tid)));
    let block_tags: Vec<Value> = block_tags.into_iter().map(|(_, _, v)| v).collect();

    // Page links — sorted by (source_id, target_id, source_page_id).
    let mut page_links: Vec<(String, String, String, Value)> = state
        .page_links
        .into_iter()
        .map(|l| {
            let sid = relabel(&l.source_id);
            let tid = relabel(&l.target_id);
            let spid = relabel_opt(&l.source_page_id);
            let spid_key = spid.as_str().unwrap_or("").to_owned();
            (
                sid.clone(),
                tid.clone(),
                spid_key,
                json!({ "source_id": sid, "target_id": tid, "source_page_id": spid }),
            )
        })
        .collect();
    page_links.sort_by(|a, b| {
        (token_key(&a.0), token_key(&a.1), token_key(&a.2)).cmp(&(
            token_key(&b.0),
            token_key(&b.1),
            token_key(&b.2),
        ))
    });
    let page_links: Vec<Value> = page_links.into_iter().map(|(_, _, _, v)| v).collect();

    // Op log digest — ordered by seq (already), filtered/canonicalized.
    let ops: Vec<Value> = state
        .op_log
        .iter()
        .map(|o| match &o.key {
            Some(k) => json!({ "op_type": o.op_type, "key": k }),
            None => json!({ "op_type": o.op_type }),
        })
        .collect();
    let op_log_digest = json!({ "count": ops.len(), "ops": ops });

    Snapshot {
        blocks,
        properties,
        block_tags,
        page_links,
        op_log_digest,
    }
}

/// Sort key for a `Bn` canonical token: `(prefix, numeric_suffix)`. Falls back
/// to lexical for non-`Bn` tokens (pass-through ids).
fn token_key(token: &str) -> (String, i64, String) {
    if let Some(rest) = token.strip_prefix('B')
        && let Ok(n) = rest.parse::<i64>()
    {
        return (String::from("B"), n, String::new());
    }
    (String::from("~"), 0, token.to_owned())
}

/// Derive `(value_type, value)` for a property row from whichever typed column
/// is set. Mirrors the TS twin exactly. Ref values are relabeled.
fn property_typed_value(p: &RawProperty, relabel: &dyn Fn(&str) -> String) -> (Value, Value) {
    if let Some(v) = &p.value_text {
        (json!("Text"), json!(v))
    } else if let Some(v) = p.value_num {
        (json!("Num"), json!(v))
    } else if let Some(v) = &p.value_date {
        (json!("Date"), json!(v))
    } else if let Some(v) = &p.value_ref {
        (json!("Ref"), json!(relabel(v)))
    } else if let Some(v) = p.value_bool {
        (json!("Bool"), json!(v))
    } else {
        (Value::Null, Value::Null)
    }
}
